### PR TITLE
Core, API: BaseReplacePartitions to branch test Refactoring

### PR DESCRIPTION
**Pull request by namrathamyske**
follow up from https://github.com/apache/iceberg/pull/5234, to refactoring minor comments left in last PR. @rdblue @amogh-jahagirdar please take a look at this.

### CODE REVIEW

**Comment by namrathamyske** [comment]
@jackye1995 can you also include this PR for the release ? This is just refactoring

**Review by jackye1995 — APPROVED** [review]
looks good to me

File: core/src/test/java/org/apache/iceberg/TestOverwrite.java [hunk 1]
```diff
@@ -182,16 +182,16 @@ public void testOverwriteFailsDelete() {
   @Test
   public void testOverwriteWithAppendOutsideOfDelete() {
     TableMetadata base = TestTables.readMetadata(TABLE_NAME);
-    long baseId =
-        latestSnapshot(base, branch) == null ? -1 : latestSnapshot(base, branch).snapshotId();
+    Snapshot latestSnapshot = latestSnapshot(base, branch);
```
Comment (amogh-jahagirdar) [inline core/src/test/java/org/apache/iceberg/TestOverwrite.java:185]: We don't need to handle in this PR but given #6651 we will end up using the SnapshotUtil api which is introduced. We can either update this PR with the proposed utility API or just later refactor to use the utility API. Since this is a test class, I'm okay either way.

**Review by amogh-jahagirdar — APPROVED** [review]
Nonblocking comment, overall LGTM!

File: core/src/test/java/org/apache/iceberg/TestOverwrite.java (hunk 1 above)
Comment (jackye1995) [reply core/src/test/java/org/apache/iceberg/TestOverwrite.java:185]: maybe we can just add the `SnapshotUtil` change in this PR? I don't think we can resolve the comments in that PR in time based on the current situation.

**Comment by jackye1995** [comment]
@namrathamyske I updated based on Amogh's comment

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java [hunk 2]
```diff
@@ -394,4 +396,42 @@ public static Schema schemaFor(Table table, Long snapshotId, Long timestampMilli
 
     return table.schema();
   }
+
+  /**
+   * Fetch the snapshot at the head of the given branch in the given table.
+   *
+   * <p>This method calls {@link Table#currentSnapshot()} instead of using branch API {@link
+   * Table#snapshot(String)} for the main branch to ensure existing tests still goes through the old
```
Comment (yyanyy) [inline core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:404]: nit/nonblocker: this class seems to be public, not sure if we want to mention the internal testing strategy here. Same goes for `latestSnapshot`

**Review by yyanyy — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java (hunk 2 above)
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:404]: good point, I can rephrase this to avoid saying tests

**Comment by jackye1995** [comment]
Merging since all comments are addressed, thanks for the review @yyanyy and @amogh-jahagirdar!
